src/exoticatechnologies/hullmods/exotics/HullmodExoticHandler.kt

```diff
@@ -0,0 +1,169 @@
+                hullmodExotic = hullmodExotic,
+                parentFleetMember = parentFleetMember
+        )
+
+        val currentInstallData = lookupMap[hullmodExoticKey]
+        if (currentInstallData == null) {
+            throw IllegalStateException("We should have had this key in the lookup map!!! lookupMap: ${lookupMap}")
+        }
+
+        val alreadyInstalledList = currentInstallData.listOfVariantsWeInstalledOn
+        // just check
+        val alreadyIn = alreadyInstalledList.contains(variant)
+        if (alreadyIn) {
+            // Actually, lets not throw just yet
+            // TODO change to throwing
+            AnonymousLogger.log("installHullmodExoticToVariant()\t\t !!!!! Trying to install on a variant that's already been installed on, returning false!!!!!", "HullmodExoticHandler")
+            return false
+        }
+        val mutableInstalledList = alreadyInstalledList.toMutableList()
+        mutableInstalledList.add(variant)
+        val newKeyValue = currentInstallData.copy(
+                parentFleetMemberAPI = currentInstallData.parentFleetMemberAPI,
+                listOfExpectedVariants = currentInstallData.listOfExpectedVariants,
+                listOfVariantsWeInstalledOn = mutableInstalledList.toList()
+        )
+        lookupMap[hullmodExoticKey] = newKeyValue
+
+        AnonymousLogger.log("<-- installHullmodExoticToVariant()\t\treturning true", "HullmodExoticHandler")
+        return true
+    }
+
+    fun removeHullmodExoticFromFleetMember(exoticHullmod: ExoticHullmod, fleetMember: FleetMemberAPI) {
+        val keysToRemove = mutableSetOf<HullmodExoticKey>()
+        val allKeys = grabAllKeysForParticularFleetMember(fleetMember)
+        for (key in allKeys) {
+            val exoticHandlerDataOptional = getDataForKey(key)
+            if (exoticHandlerDataOptional.isPresent()) {
+                val exoticHandlerData = exoticHandlerDataOptional.get()
+                for (variant in exoticHandlerData.listOfVariantsWeInstalledOn) {
+                    val variantHullSize = variant.hullSpec.hullSize
+                    exoticHullmod.removeEffectsBeforeShipCreation(variantHullSize, variant.statsForOpCosts, exoticHullmod.hullModId)
+                    // Lets keep track of 'removed' keys here because why not
+                    keysToRemove.add(key)
+                }
+            }
+        }
+        // Log before
+        AnonymousLogger.log("removeHullmodExoticFromFleetMember()\tlookupMap before cleaning: ${lookupMap}", "HullmodExoticHandler")
+
+        // Now, cleanup the lookupMap from all the to-remove keys
+        for(key in keysToRemove) {
+            lookupMap.remove(key)
+        }
+        // Log after
+        AnonymousLogger.log("<-- removeHullmodExoticFromFleetMember()\tlookupMap AFTER cleaning: ${lookupMap}", "HullmodExoticHandler")
+    }
+
+    private fun grabAllKeysForParticularFleetMember(fleetMemberAPI: FleetMemberAPI): List<HullmodExoticKey> {
```

Comment:
add javadoc even though it's a private method. just say what it's for and what it's used for.

Reply:
all comments relating to `HullmodExoticHandler` are **most likely** either outdated or done.

Reply:
resolving.